File: Momchil psychopy code/context.py
from psychopy import core, visual, gui, data, event
from psychopy.tools.filetools import fromFile, toFile
import random
import datetime
import os
import string

#
# Constants
#

nTest = 4
nTrain = 20
trial = -1 # current trial
groups = [1,2,3]  # subject group(s)
random.shuffle(groups)
m = groups[0] # pick one TODO do all

reward = 0
mode = 1
restaurants = ["Molina's Cantina", "Restaurante Arroyo", "El Coyote Cafe"]
foods = ["food1.png", "food2.png","food3.png"]
train = [0,0,0,0,0,1,1,1,1,1,2,2,2,2,2,3,3,3,3,3] # sequence of training trials
random.shuffle(train)
test = [0,1,2,3] # sequence of test trials
random.shuffle(test)
food = 0
context = 0
Test = 0 # are we in the testing phase?
f = 0

now = datetime.datetime.now()
subjId = ''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(10))
filename = subjId + '_' + now.strftime('%Y-%m-%d_%H-%M-%S') + '_' + str(groups[0]) + str(groups[1]) + str(groups[2]) + '.csv'
print 'filename = ', filename
filename = os.path.join('data', filename)
dataFile = open(filename, 'w') # TODO bufsize 0 to flush always?

# outcomes -- did the customer get sick on train[trial] (not defined for test trials)
# 1 = sick, 0 = not sick
#
r = None
if m == 1:
    r = [1, 0, 1, 0]
elif m == 2:
    r = [1, 0, 0, 1]
else:
    assert m == 3
    r = [1, 1, 0, 0]

train_cue = [0, 1, 0, 1] # training foods for each trial type
train_context = [0, 0, 1, 1] # training restaurant for each trial type
test_cue = [0, 0, 2, 2] # testing foods for each trial type
test_context = [0, 2, 0, 2] # testing restaurant for each trial type

# TODO

# record responses

# have diff conditions in each subject
# cross-balance based on subject id (6 perms, use mod 6)
# have different foods & restaurant names (french? italian?)

# better UI -- correct / wrong
# style it -- colors, etc
# instructions / extra messages & boilerplate from Sam JS script

# QUESTIONS

# ipython/cmd line -- how to?



#
# Create window and stimuli
#

win = visual.Window([1000,800],allowGUI=True, monitor='testMonitor', units='deg')
restaurant_txt = visual.TextStim(win, pos=[0, +6.5], text="Restaurant name")
food_img = visual.ImageStim(win, "food1.png", pos=[0, +3])
sick_img = visual.ImageStim(win, "sick.png")
sick_txt = visual.TextStim(win, pos=[-6,-9],text='Sick\n<-') # sick = left
notsick_img = visual.ImageStim(win, "smiley.png")
notsick_txt = visual.TextStim(win, pos=[+6,-9],
    text="Not sick\n->") # not sick = right
feedback_txt = visual.TextStim(win, pos=[0, -9])
correct_txt = visual.TextStim(win, pos=[0, -1.5], text="CORRECT", color='blue', bold=True)
wrong_txt = visual.TextStim(win, pos=[0, -1.5], text="WRONG", color='red', bold=True)
#and some handy clocks to keep track of time
globalClock = core.Clock()

#
# Run experiment
#

for _ in range(nTrain + nTest):
    trial += 1
    
    # get restaurant & food on trial X
    #
    if Test == 0 and trial == nTrain:
        Test = 1
        trial = 0

    if Test == 1:
        cue = test_cue[test[trial]]
        context = test_context[test[trial]]
    else:
        cue = train_cue[train[trial]]
        context = train_context[train[trial]]
    food = foods[cue]
    restaurant = restaurants[context]
    
    print _, ' -- trial:', trial, 'Test:', Test, 'food:', food, 'restaurant', restaurant
    
    # UI setup
    #
    food_img.setImage(food)
    restaurant_txt.setText(restaurant)
    sick_img.pos = [-6, -5];
    notsick_img.pos = [+6, -5];
    
    # show restaurant & food
    #
    food_img.draw()
    restaurant_txt.draw()
    sick_img.draw()
    sick_txt.draw()
    notsick_img.draw()
    notsick_txt.draw()
    win.flip()
    #core.wait(1)

    # get user response
    #
    response = None
    while response is None: # keep trying until user presses left or right
        allKeys = event.waitKeys()
        print '          key = ', allKeys
        for thisKey in allKeys:
            if thisKey=='left': # sick = left
                response = 1
            elif thisKey=='right':
                response = 0
            elif thisKey in ['q', 'escape']:
                core.quit() #abort experiment
    assert response is not None
   
    # give feedback
    #
    if not Test:
        outcome = r[train[trial]]
        if outcome:
            sick_img.pos = [0, -5];
            feedback_txt.setText("The customer got sick!\nPress any key to continue")
            sick_img.draw()
        else:
            notsick_img.pos = [0, -5];
            feedback_txt.setText("The customer didn't get sick!\nPress any key to continue")
            notsick_img.draw()
        food_img.draw()
        restaurant_txt.draw()
        feedback_txt.draw()
        if outcome == response:
            correct_txt.draw()
        else:
            wrong_txt.draw()
        win.flip()
        
        allKeys = event.waitKeys()
    else:
        outcome = -1 # no "known" outcome in test case
        
    dataFile.write("%d,%d,%d,%d\n" % (outcome, context, cue, response))
        
    
    

File: Sam JS code/context.js
var nTest = 4;
var nTrain = 20;
var trial = -1;
var d = new Date();
var m = [1,2,3]; m = shuffle(m); m = m[0];
var subjID = '7' + Math.random().toString().substring(3,8);
var filename = 'data/' + subjID + '_' + d.getTime() + '_' + m +  '.csv';
var reward = 0;
var mode = 1;
var restaurants = ["Molina's Cantina", "Restaurante Arroyo", "El Coyote Cafe"];
var foods = ["food1", "food2","food3"];
var train = [0,0,0,0,0,1,1,1,1,1,2,2,2,2,2,3,3,3,3,3]; train = shuffle(train);
var test = [0,1,2,3]; test = shuffle(test);
var food = 0;
var context = 0;
var Test = 0;
var f = 0;

var r = [];
if (m==1) {
	r[0] = 1;
	r[1] = 0;
	r[3] = 1;
	r[4] = 0;
} else if (m==2) {
	r[0] = 1;
	r[1] = 0;
	r[3] = 0;
	r[4] = 1;
} else {
	r[0] = 1;
	r[1] = 1;
	r[3] = 0;
	r[4] = 0;
}

var train_cue = [];
train_cue[0] = 0;
train_cue[1] = 1;
train_cue[2] = 0;
train_cue[3] = 1;

var train_context = [];
train_context[0] = 0;
train_context[1] = 0;
train_context[2] = 1;
train_context[3] = 1;

var test_cue = [];
test_cue[0] = 0;
test_cue[1] = 0;
test_cue[2] = 2;
test_cue[3] = 2;

var test_context = [];
test_context[0] = 0;
test_context[1] = 2;
test_context[2] = 0;
test_context[3] = 2;

// Initialization
$(document).ready(function() {
 	$('#endExperiment').hide();
	$('#startGame').hide();
 	$('#buttons').hide();
	$('#food1').hide();
	$('#Instructions').hide();
	$('#feedback').hide();
	$('#submit').hide();
	$('#food1').hide();
	$('#food2').hide();
	$('#food3').hide();
	$('#smiley').hide();
	$('#sick').hide();
	$('#restaurant').hide();
	$('#button1').text("Sick");
	$('#button2').text("Not sick");
	$('#submit').text("Next");
				  
    $("#button1").click(function() {
		b = 1;
		Feedback();
	})
  
    $("#button2").click(function() {
		b = 0;
		Feedback();
	})
				  
	$("#submit").click(function() {
		NextTrial();
	})

});

function Feedback() {
	
	if (Test==0) {
		$('#button1').hide();
		$('#button2').hide();
		$('#submit').show();
		if (r[train[trial]]==1) {
			var outcome = 1;
			$('#sick').show();
			$('#title').text("The customer got sick!");
		} else {
			var outcome = 0;
			$('#smiley').show();
			$('#title').text("The customer didn't get sick!");
		}
		var result_string = outcome + ',' + context + ',' + f + ',' + b + '\n';
		$.post("post_results.php",{postresult: result_string, postfile: filename});
	} else {
		var outcome = -1;
		var result_string = outcome + ',' + context + ',' + f + ',' + b + '\n';
		$.post("post_results.php",{postresult: result_string, postfile: filename});
		NextTrial();
	}
	
}

function NextTrial() {
	trial++;
	$('#title').text("Predict whether the customer will get sick from this food.");
	$('#'+food).hide();
	$('#smiley').hide();
	$('#sick').hide();
	$('#button1').show();
	$('#button2').show();
	$('#submit').hide();
	
	if (Test==0 && trial==nTrain) {
		Test = 1;
		trial = 0;
	}
	
	if (Test==1) {
		f = test_cue[test[trial]];
		food = foods[f];
		context = test_context[test[trial]];
	} else {
		f = train_cue[train[trial]];
		food = foods[f];
		context = train_context[train[trial]];
	}
	
	$('#restaurant').text(restaurants[context]);
	$('#restaurant').show();
	$('#'+food).show();
	
	// end of experiment
	if (Test==1 && trial==nTest) {
		$('#buttons').hide();
		$('#startGame').hide();
		$('#Instructions').show();
		$('#title').hide();
		$('#restaurant').hide();
		$('#'+food).hide();
		$("#Instructions").text("You're done! \nYour code is " + subjID + ". Please return to the Mechanical Turk page to enter this code and get paid.");
	}
}

function StartGame() {
	$('#startGame').hide();
	$('#Instructions').hide();
	$('#buttons').show();
	NextTrial();
}

function StartExperiment() {
	$('#consent').hide();
	$('#startExperiment').hide();
	$('#Instructions').show();
	$('#startGame').show();
}

function shuffle(array) {
    var currentIndex = array.length
    , temporaryValue
    , randomIndex
    ;
    
    // While there remain elements to shuffle...
    while (0 !== currentIndex) {
        
        // Pick a remaining element...
        randomIndex = Math.floor(Math.random() * currentIndex);
        currentIndex -= 1;
        
        // And swap it with the current element.
        temporaryValue = array[currentIndex];
        array[currentIndex] = array[randomIndex];
        array[randomIndex] = temporaryValue;
    }
    
    return array;
}